daemons/attrd/attrd_commands.c

```diff
@@ -617,7 +618,8 @@ attrd_peer_message(crm_node_t *peer, xmlNode *xml)
             clear_attribute_value_seen();
         }
 
-        for (child = __xml_first_child(xml); child != NULL; child = __xml_next(child)) {
+        for (child = pcmk__xml_first_child(xml); child != NULL;
+             child = pcmk__xml_next(child)) {
```

Comment:
We use this pattern so much.  It might be worth a separate PR that introduces something like `pcmk__xml_child_foreach` that takes a parent and a callback function, and the callback function type looks like you'd expect.

Reply:
It would be cleaner, but I'd be concerned about the function call overhead (these are static inline functions currently) since XML processing happens a whole lot. Some timing comparisons might help decide.

Reply:
I think it'd be worth looking into.  If you have any good pointers on how to do useful timing comparisons, I'd be happy to look at it... eventually.

Reply:
Taking the largest cts/scheduler/*.xml file and running crm_simulate with a big --repeat would probably be a decent approach. Or creating a test CIB with 1000 nodes and 1000 resources and a short test program with nested loops through those. It's probably insignificant, but I'd want to be sure.

We have several related functions:
* pcmk__xml_first_child()/pcmk__xml_next() return XML nodes of any kind other than text (document, element, comment, CDATA, ...)
* pcmk__xe_first_child()/pcmk__xe_next() return just XML elements
* first_named_child()/crm_next_same_xml() return just XML elements of a given type

Currently we don't always use the proper pair for a given context, so that should be part of any overhaul. E.g. we almost certainly have places where it would choke on a comment because it's using the first pair when it should be using one of the others. And we have lots of places that could use the last pair but are string-comparing the element type themselves (for no good reason, they were probably just written before those existed).